hostmap.go

```diff
@@ -429,29 +432,83 @@ func (hm *HostMap) DeleteRelayIdx(localIdx uint32) {
 	delete(hm.RemoteIndexes, localIdx)
 }
 
+func (hm *HostMap) MakePrimary(hostinfo *HostInfo) {
+	hm.Lock()
+	defer hm.Unlock()
+	hm.unlockedMakePrimary(hostinfo)
+}
+
+func (hm *HostMap) unlockedMakePrimary(hostinfo *HostInfo) {
+	//TODO: should we try and coalesce the relay state from oldHostinfo here?
```

Comment:
I think the relays should be tied to a single HostInfo object, and not migrated from one HostInfo instance to another.

Relay packets are authenticated with the keys stored in the HostInfo ConnectionState, and each HostInfo for the same vpn IP will have different encryption/decryption keys. Knowing which keys to use, and when it's OK to switch from one HostInfo's keys to another would be tricky:tm:.

Relay state is also shared between HostInfo objects on different peers. Migrating RelayState from one HostInfo object instance to another HostInfo instance on one peer implies that the HostInfo on the remote peer also has the corresponding RelayState. However, it's possible that a handshake that resulted in 2 HostInfo objects on _this_ peer is due to the _remote_ peer restarting its Nebula process, losing its initial HostInfo state. In that case, _this_ peer would migrate the Relay state that doesn't actually exist on the _remote_ peer.

I believe the best way to detect this kind of divergence is by way of the ConnectionState keys. As long as the remote peer still has the ConnectionState keys, _this_ peer can reliably conclude that the _remote_ peer also still has associated RelayState info, and vice versa.

Reply:
Will be handled in a very near future PR. The current failure scenarios match the existing failure scenarios in master to the best of my knowledge/testing.